src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/handlers/VideoBookingAmendedEventHandler.kt

```diff
@@ -45,43 +47,90 @@ class VideoBookingAmendedEventHandler(
               }
             }
 
-          if (activitiesService.isAppointmentsRolledOutAt(booking.prisonCode())) {
-            val appointmentTypesForPrisoner = (history.appointments().map { it.prisonerNumber to it.appointmentType } + booking.appointments().map { it.prisonerNumber to it.appointmentType }).toSet()
-            appointmentTypesForPrisoner.forEach { (prisonerNumber, type) ->
-              val oldAppointment = history.appointments().singleOrNull { it.appointmentType == type && it.prisonerNumber == prisonerNumber }
-              val newAppointment = booking.appointments().singleOrNull { it.appointmentType == type && it.prisonerNumber == prisonerNumber }
+          if (activitiesService.isAppointmentsRolledOutAt(amendedBooking.prisonCode())) {
+            when (amendedBooking.bookingType) {
+              BookingType.COURT -> {
+                val (previousPreHearing, amendedPreHearing) = oldBooking.preHearing() to amendedBooking.preHearing()
+                val (previousMainHearing, amendedMainHearing) = oldBooking.mainHearing() to amendedBooking.mainHearing()!!
+                val (previousPostHearing, amendedPostHearing) = oldBooking.postHearing() to amendedBooking.postHearing()
 
-              when {
-                oldAppointment == null -> {
-                  manageExternalAppointmentsService.createAppointment(newAppointment!!)
-                }
-
-                newAppointment == null -> {
-                  manageExternalAppointmentsService.cancelPreviousAppointment(oldAppointment)
-                }
+                // All cancellations must be done first
+                cancelIfRemoved(previousPreHearing, amendedPreHearing)
+                cancelIfRemoved(previousPostHearing, amendedPostHearing)
 
-                else -> {
-                  manageExternalAppointmentsService.amendAppointment(oldAppointment, newAppointment)
+                if (amendedMainHearing.isTheSameAs(previousMainHearing)) {
```

Comment:
I think this logic is right.  Can talk through if not clear enough and tweak if need be.